### ISSUE TITLE
[Bug]: A3 ChunkKdaFwd varlen 路径输出 bitwise 不确定

### ISSUE THREAD

**Issue by ChenxiQ** [flashserve/flash-linear-attention-npu#440]
### 环境信息

- CANN 完整版本: 9.0.0
- Python: 3.11.15
- PyTorch: 2.10.0+cpu
- torch_npu: 2.10.0.post2
- fla_npu: 26.7.0.dev0，main commit `a333136`（2026-08-31 快照，`FLA_NPU_SOC=ascend910_93` 全量构建，OPP 内嵌 wheel）
- SOC / 设备型号: ascend910_93 / 910C
- 安装方式: 一体化 wheel，OPP 内嵌于 wheel
- custom OPP 环境状态: wheel 内置 OPP 已命中
- fla_npu runtime readiness: `chunk_kda_fwd` 可导入并执行

### 最小复现

```python
单文件脚本（零外部依赖，仅需 fla_npu），source CANN 环境后运行：


source /usr/local/Ascend/ascend-toolkit/set_env.sh
ASCEND_RT_VISIBLE_DEVICES=0 python3 repro_chunk_kda_fwd_nondet.py



# repro_chunk_kda_fwd_nondet.py
import torch
import torch_npu  # noqa: F401
from fla_npu.ops.triton import l2norm_fwd
from fla_npu.ops.ascendc import chunk_kda_fwd

H, K, V = 2, 128, 128
T, SEGS, ITERS = 2048, 10, 20
fails = 0
for it in range(ITERS):
    torch.manual_seed(it)
    q = torch.randn(1, T, H, K, dtype=torch.bfloat16, device="npu")
    k = torch.randn_like(q)
    v = torch.randn(1, T, H, V, dtype=torch.bfloat16, device="npu")
    g = torch.randn(1, T, H, K, dtype=torch.bfloat16, device="npu")
    beta = torch.randn(1, T, H, dtype=torch.float32, device="npu").sigmoid()
    A_log = torch.randn(H, dtype=torch.float32, device="npu")
    dt_bias = torch.randn(H * K, dtype=torch.float32, device="npu")
    # 注意: l2norm 必须做（与模型调用一致）；未归一化的 q/k 会因 WY 表示
    # 数值爆炸产生"合法 NaN"，污染确定性判据
    q, _ = l2norm_fwd(q)
    k, _ = l2norm_fwd(k)
    cuts = sorted(torch.randperm(T - 1)[: SEGS - 1].tolist())
    cu = [0] + [c + 1 for c in cuts] + [T]          # varlen: 10 段随机切分

    def call():
        return chunk_kda_fwd(
            q, k, v, g, beta, K ** -0.5, 64,
            layout="BSND", output_final_state=True, cu_seqlens=cu,
            safe_gate=True, lower_bound=-5.0, use_gate_in_kernel=True,
            A_log=A_log, dt_bias=dt_bias, disable_recompute=False,
            return_intermediate_states=False, state_v_first=False,
        )

    o1 = call()
    ref = [t.clone() if t is not None else None for t in o1]
    o2 = call()
    torch.npu.synchronize()
    for j, (a, b) in enumerate(zip(ref, o2)):
        if a is None or b is None:
            continue
        # NaN 感知的逐位比较
        same = (torch.equal(a.view(torch.int16), b.view(torch.int16))
                if a.dtype == torch.bfloat16
                else torch.equal(a.view(torch.int32), b.view(torch.int32)))
        if not same:
            print(f"[NONDET] iter={it} output[{j}] bitwise differs")
            fails += 1

print(f"RESULT failures={fails}/{ITERS} -> "
      f"{'NONDETERMINISTIC' if fails else 'DETERMINISTIC'}")
```

### 完整报错

```text
实测输出（910_93 单卡）：


[NONDET] iter=0 output[1] bitwise differs
[NONDET] iter=5 output[0] bitwise differs
...（共 11/20 轮出现）
RESULT failures=11/20 -> NONDETERMINISTIC


把 `cu = ...` 改为 `cu = None`（dense）则输出 `RESULT failures=0/20 -> DETERMINISTIC`（已实测）。
```

### 期望行为

相同输入、相同属性和相同执行环境下重复调用 `chunk_kda_fwd`，全部返回位应保持 bitwise deterministic；A3 的 varlen 路径（含跨 chunk 尾块与多段边界）不应出现输出波动。

## 触发条件矩阵

| 配置 | 结果 |
| --- | --- |
| 融合 chunk_kda_fwd + varlen（T=131072/30 段，dr=False） | **21/50 迭代逐位不一致** |
| 融合 chunk_kda_fwd + varlen（T=2048/10 段） | **11/20** |
| 融合 chunk_kda_fwd + varlen（dr=True，全中间量导出） | 28 处失败/20 轮 |
| 融合 chunk_kda_fwd + **dense** | **0/20，完全确定** |
| fla-org triton chunk_kda_fwd + varlen（同输入同条件） | 0/40，完全确定 |
| 独立 ChunkGatedDeltaRuleFwdH + varlen（用融合算子导出的 kg/w/u/gk 作输入，双调用） | 0/60，完全确定 |

差异幅度为 bf16 舍入级（max_finite_abs 约 1e-06 ~ 9e-04），输出全部有限；输入张量从未被原地改写；调用前后放置的 canary 守护张量从未被破坏；T 从 2048 到 131072 均可触发，与对齐无关。

## 输出分布（阶段定位）

`disable_recompute=True` 下对全部 12 个输出做双调用对比（T=131072 / 30 段，20 轮共 28 处失败）：

| 输出 | 抖动次数 | 所属阶段 |
| --- | ---: | --- |
| final_state | 25 | FwdH/Finalize（跨 chunk 状态递推终点） |
| attn_out | 16 | Finalize |
| w | 15 | PostWu |
| v_new | 5 | FwdH |
| gk / Aqk / Akk / u / qg / kg / h | **0** | Prepare / Gate / PostWu / FwdH |

抖动集中在跨 chunk 状态递推链（FwdH 阶段）及其上下游产物；chunk 内计算（Prepare/Gate）全部稳定。独立调用 FwdH 算子（同中间量、同 varlen）完全确定——即"单 kernel 正常、融合链路抖动"。

## 与 SFT 训练崩溃的关联

在 4 机×16 卡 910C 的 SFT 训练（131k pack varlen、SP8、gradient checkpointing `use_reentrant=False`）中启用该融合前向（反向仍走 triton）：

- 首个 backward 触发 `torch.utils.checkpoint.CheckpointError`：重算前向的 token 数与原始前向差 ±1（方向随机，仅个别 rank 命中）；
- A/B 消融：完全相同的接入代码，仅把前向从本算子换回 triton 即恢复干净（两轮对照均零报错，loss 逐位一致）——崩溃严格随本算子的调用出现/消失；
- CheckpointError 只校验张量元数据（shape/dtype），数值抖动本身不足以触发；推测是同一竞态在训练的内存布局/多流时序下恶化为越界写、破坏了相邻张量（该机制未在单算子环境单独复现）。

## 其他补充

- 诊断建议对照 `KDA_FORWARD_LESSONS.md` §9.4（L0 读依赖未注册导致 workspace 提前复用——"单跑正常、拼接后错误"）与 §10.4（跨 pipe event / cross-core flag / UB slot 复用 / GM workspace 写区重叠嫌疑清单）；
- varlen 触发、dense 干净的边界条件可作为定位线索：varlen 的 per-sequence chunk 划分与尾块处理改变了核间任务划分与同步配对；
- #428（补齐 chunk_kda_fwd 确定性测试交付）落实时建议把本复现纳入 A3 矩阵。

**Comment by weinachuan** [comment]
已完成定位、修复和版本回合，结论如下。

### 根因

变长序列尾块的 PostWU 路径会复用同一段 GM：`preparedQG` 作为 `w_seed` 输入，同时 `w` 又写回该地址。旧实现将两个 AIV 按输出行分工，并从低行向高行原地写回；但每个 `w` 输出行都需要根据下三角 `Akk` 读取多个尚存活的 `w_seed` 行。提前写回会覆盖其他行或另一个 AIV 后续仍要读取的源数据。

现有 `SetFlag/WaitFlag` 只约束单个 AIV 内部的 MTE/VEC 流水，不能建立两个 AIV 之间的执行顺序，因此不同调度下可能读到覆盖前或覆盖后的值，最终表现为 `w` bitwise 不确定。`w` 随后进入 FwdH，可能继续影响 `v_new`、`final_state` 和少量 `attn_out`。

该问题只在 `curT < chunk_size` 的变长尾块 Vector 路径触发；完整块通过 AIC scratch 和核间同步交接，所以 dense 和 64 对齐变长场景保持稳定。

### 修复

- 将 `w` 的两个 AIV 任务改为按 K 维列区间分工，保证写入区间互不重叠。
- 按行从高到低计算和写回，保证下三角依赖的每个 `w_seed` 行都在最后一次读取后才被覆盖。
- `u_seed` 与 `u` 不存在地址复用，继续按行并行。
- generic（A2/A3）和 arch35（A5）路径同步修复，并增加 63-token 尾块、全部 11 个可见输出的确定性回归。

### 验证结果

- A2 原始 T=2048、10 段变长用例：未修基线 20 轮中有 2 轮漂移，修复后 20/20 稳定；最终提交的放大尾块用例 100/100 稳定，dense 与 64 对齐变长对照也保持稳定。
- A5 相同尾块回归：未修基线后续 19/19 轮的 `w` 均变化，修复后 100 轮全部 11 个输出逐位一致且 finite。
- A3 单算子目标编译通过。
- 主线和 `v26.9.0` 的 NPU CI、精度检查均通过。

修复已合入：

- 主线：[PR #441](https://github.com/flashserve/flash-linear-attention-npu/pull/441)
- `v26.9.0`：[PR #442](https://github.com/flashserve/flash-linear-attention-npu/pull/442)

本次结论针对 #440 的 bitwise 确定性问题；额外边界与精度观察已在对应 PR 描述中单独记录，未作为本问题的通过依据。
